grid_map_ros/include/grid_map_ros/Costmap2DConverter.hpp

```diff
@@ -0,0 +1,314 @@
+  {
+    // Check compliance.
+    Size size(costmap2d.getSizeInCellsX(), costmap2d.getSizeInCellsY());
+    if ((outputMap.getSize() != size).any()) {
+      ROS_ERROR("Costmap2D and output map have different sizes!");
+      return false;
+    }
+    if (!outputMap.getStartIndex().isZero()) {
+      ROS_ERROR("CostmapConverter::fromCostmap2D() does not support non-zero start indices!");
+      return false;
+    }
+    // Copy data.
+    // Reverse iteration is required because of different conventions
+    // between Costmap2D and grid map.
+    typename MapType::Matrix data(size(0), size(1));
+    const size_t nCells = outputMap.getSize().prod();
+    for (size_t i = 0, j = nCells - 1; i < nCells; ++i, --j) {
+      const unsigned char cost = costmap2d.getCharMap()[j];
+      data(i) = (float) costTranslationTable_[cost];
+    }
+
+    outputMap.add(layer, data);
+    return true;
+  }
+
+  /*!
+   * Load the costmap2d into the specified layer.
+   * @warning This does not lock the costmap2d object, you should take care to do so from outside this function.
+   * @param costmap2d from this costmap type object.
+   * @param layer the name of the layer to insert.
+   * @param outputMap to this costmap type object.
+   * @return true if successful, false otherwise.
+   */
+  bool addLayerFromCostmap2D(costmap_2d::Costmap2DROS& costmap2d, const std::string& layer,
+                             MapType& outputMap)
+  {
+    return addLayerFromCostmap2D(*(costmap2d.getCostmap()), layer, outputMap);
+  }
+
+  /*!
+   * Initialize the output map at the robot pose of the underlying Costmap2DROS object.
+   *
+   * Note: This needs some beyond just fixing the center of the output map to the robot pose
+   * itself. To avoid introducing error, you want to shift the center so the
+   * underlying costmap2d cells align with the newly created map object. This does
+   * that here.
+   *
+   * @warning this does not lock the costmap2d object, you should take care to do so from
+   * outside this function in scope with any addLayerFromCostmap2DAtRobotPose calls you wish to make.
+   *
+   * @param costmap2d : the underlying Costmap2DROS object
+   * @param geometry: size of the subwindow to snapshot around the robot pose
+   * @param outputMap : initialise the output map with the required parameters
+   */
+  bool initializeFromCostmap2DAtRobotPose(costmap_2d::Costmap2DROS& costmap2d, const Length& length,
+                                          MapType& outputMap)
+  {
+    const double resolution = costmap2d.getCostmap()->getResolution();
+
+    // Get the Robot Pose Transform.
+    tf::Stamped<tf::Pose> tfPose;
+    if(!costmap2d.getRobotPose(tfPose))
+    {
+      std::ostringstream error_message;
+      error_message << "Could not get robot pose, is it actually published?";
+      ROS_ERROR_STREAM("CostmapConverter:: " << error_message.str());
+      return false;
+      // throw std::runtime_error(error_message.str());
```

Comment:
I'm happy to use exceptions instead of the `ROS_ERROR` and returning `false`. Is there a common sense when to use which?